synapse/federation/sender/__init__.py

```diff
@@ -560,3 +577,37 @@ async def get_replication_rows(
         # Dummy implementation for case where federation sender isn't offloaded
         # to a worker.
         return [], 0, False
+
+    async def _wake_destinations_needing_catchup(self):
+        """
+        Wakes up destinations that need catch-up and are not currently being
+        backed off from.
+
+        In order to reduce load spikes, adds a delay between each destination.
+        """
+
+        last_processed = None  # type: Optional[str]
+
+        while True:
+            destinations_to_wake = await self.store.get_catch_up_outstanding_destinations(
+                last_processed
+            )
+
+            if not destinations_to_wake:
+                # finished waking all destinations!
+                self._catchup_after_startup_timer = None
+                break
+
+            destinations_to_wake = [
+                d
+                for d in destinations_to_wake
+                if self._federation_shard_config.should_handle(self._instance_name, d)
+            ]
+
+            for destination in destinations_to_wake:
+                last_processed = destination
+                logger.info(
+                    "Destination %s has outstanding catch-up, waking up.", destination,
+                )
+                self.wake_destination(destination)
+                await self.clock.sleep(CATCH_UP_STARTUP_INTERVAL_SEC)
```

Comment:
Kind of just a style thing, but I think this is equivalent. Don't need to make this change though if you think the other way is clearer.

```suggestion
            for last_processed in destinations_to_wake:
                logger.info(
                    "Destination %s has outstanding catch-up, waking up.", last_processed,
                )
                self.wake_destination(last_processed)
                await self.clock.sleep(CATCH_UP_STARTUP_INTERVAL_SEC)
```